src/Nethermind/Nethermind.Merge.AuRa/Shutter/ShutterTxSource.cs

```diff
@@ -0,0 +1,333 @@
+        return [];
+    }
+
+    public void OnDecryptionKeysReceived(Dto.DecryptionKeys decryptionKeys)
+    {
+        if (decryptionKeys.Gnosis.Slot <= _loadedTransactionsSlot)
+        {
+            if (_logger.IsDebug) _logger.Debug($"Skipping Shutter decryption keys from slot {decryptionKeys.Gnosis.Slot}, keys currently stored for slot {_loadedTransactionsSlot}.");
+            return;
+        }
+
+        ShutterEon.Info? eonInfo = _eon.GetCurrentEonInfo();
+        if (eonInfo is null)
+        {
+            if (_logger.IsDebug) _logger.Debug("Cannot check Shutter decryption keys, eon info was not found.");
+            return;
+        }
+
+        if (_logger.IsDebug) _logger.Debug($"Checking Shutter decryption keys instanceID: {decryptionKeys.InstanceID} eon: {decryptionKeys.Eon} #keys: {decryptionKeys.Keys.Count} #sig: {decryptionKeys.Gnosis.Signatures.Count()} #txpointer: {decryptionKeys.Gnosis.TxPointer} #slot: {decryptionKeys.Gnosis.Slot}");
+
+        if (CheckDecryptionKeys(decryptionKeys, eonInfo.Value))
+        {
+            if (_logger.IsInfo) _logger.Info($"Validated Shutter decryption key for slot {decryptionKeys.Gnosis.Slot}");
+
+            List<SequencedTransaction> sequencedTransactions = GetNextTransactions(decryptionKeys.Eon, decryptionKeys.Gnosis.TxPointer);
+            if (_logger.IsInfo) _logger.Info($"Got {sequencedTransactions.Count} transactions from Shutter mempool...");
+
+            _loadedTransactions = DecryptSequencedTransactions(sequencedTransactions, decryptionKeys);
+            _loadedTransactionsSlot = decryptionKeys.Gnosis.Slot;
+
+            if (_logger.IsInfo)
+            {
+                _loadedTransactions.ForEach(tx => _logger.Info(tx.ToShortString()));
+            }
+        }
+    }
+
+    private ulong GetNextSlot()
+    {
+        // assume Gnosis or Chiado chain
+        ulong genesisTimestamp = (_specProvider.ChainId == BlockchainIds.Chiado) ? ChiadoSpecProvider.BeaconChainGenesisTimestamp : GnosisSpecProvider.BeaconChainGenesisTimestamp;
+        return (((ulong)DateTimeOffset.UtcNow.ToUnixTimeSeconds() - genesisTimestamp) / 5) + 1;
+    }
+
+    private IEnumerable<Transaction> DecryptSequencedTransactions(IEnumerable<SequencedTransaction> sequencedTransactions, Dto.DecryptionKeys decryptionKeys)
+    {
+        // order by identity preimage to match decryption keys
+        IEnumerable<(int, Transaction?)> unorderedTransactions = sequencedTransactions
+            .Select((x, index) => x with { Index = index })
+            .OrderBy(x => x.IdentityPreimage, Bytes.Comparer)
+            .Zip(decryptionKeys.Keys.Skip(1))
+            .Select(x => (x.Item1.Index, DecryptSequencedTransaction(x.Item1, x.Item2)));
+
+        // return decrypted transactions to original order
+        IEnumerable<Transaction> transactions = unorderedTransactions
+            .OrderBy(x => x.Item1)
+            .Select(x => x.Item2)
+            .OfType<Transaction>();
```

Comment:
Do we need all this LINQ?

Reply:
What's wrong with LINQ?

Reply:
Speed sucks, double enumeration is easy to do - you do it a lot.

Reply:
In this case I'm not sure how to optimise. The `Select` statements could be replaced with foreach, but it is bottlenecked by having to sort between each `Select` so I don't think it would make a difference. Also it seems quite readable with LINQ